File: go/kindergarten-garden/kindergarten_garden.go
package kindergarten

import (
	"errors"
	"sort"
	"strings"
)

type Garden map[string][]string

var plantDict = map[string]string{"V": "violets", "R": "radishes", "G": "grass", "C": "clover"}

func NewGarden(diagram string, children []string) (*Garden, error) {
	g := Garden{}
	rows := strings.Split(strings.TrimSpace(diagram), "\n")
	if len(rows[0]) != len(rows[1]) || string(diagram[0]) != "\n" || len(rows[0])%2 != 0 {
		return &g, errors.New("")
	}
	kids := append(make([]string, 0, len(children)), children...)
	sort.Strings(kids)
	for i, child := range kids {
		_, exists := g[child]
		if exists {
			return &g, errors.New("child already has plants")
		}
		base := 2 * i
		g[child] = make([]string, 0, 4)
		for j := range rows {
			val1, exists1 := plantDict[string(rows[j][base])]
			val2, exists2 := plantDict[string(rows[j][base+1])]
			if !exists1 || !exists2 {
				return &g, errors.New("invalid plant abbreviation")
			}
			g[child] = append(g[child], val1, val2)
		}
	}
	return &g, nil
}
func (g *Garden) Plants(child string) (row []string, exists bool) {
	row, exists = (*g)[child]
	return
}

File: c/kindergarten-garden/kindergarten_garden.c
#include "kindergarten_garden.h"
#include <string.h>

static plant_t char_to_plant(char c) {
    switch (c) {
        case 'V':
            return VIOLETS;
        case 'G':
            return GRASS;
        case 'C':
            return CLOVER;
        default:
            return RADISHES;
    }
}

plants_t plants(const char *diagram, const char *student) {
    plants_t res;
    long row_len = strchr(diagram, '\n') - diagram;
    int base_idx = student[0] - 'A';
    res.plants[0] = char_to_plant(diagram[base_idx * 2]);
    res.plants[1] = char_to_plant(diagram[base_idx * 2 + 1]);
    res.plants[2] = char_to_plant(diagram[row_len + base_idx * 2 + 1]);
    res.plants[3] = char_to_plant(diagram[row_len + base_idx * 2 + 2]);
    return res;
}
